core/modules/loader/CentralWorker.cc

```diff
@@ -0,0 +1,1114 @@
+            ":udp=" << _udpPort << " tcp=" << _tcpPort << ")";
+    return os.str();
+}
+
+void CentralWorker::_startMonitoring() {
+    // Add _workerList to _doList so it starts checking new entries.
+    _centralWorkerDoListItem = std::make_shared<CentralWorkerDoListItem>(this);
+    doList->addItem(_wWorkerList);
+    doList->addItem(_centralWorkerDoListItem);
+}
+
+
+void CentralWorker::_monitor() {
+    LOGS(_log, LOG_LVL_INFO, "CentralWorker::_monitor");
+
+    /// If our id is invalid, try registering with the master.
+    if (_isOurIdInvalid()) {
+        _registerWithMaster();
+        // Give the master a half second to answer.
+        usleep(500000);
+        return;
+    }
+
+    // If data gets shifted, check everything again as ranges will have
+    // changed and there may be a lot more data to shift.
+    bool dataShifted = false;
+    do {
+        // TODO Check if we've heard from left neighbor (possibly kill connection if nothing heard???)
+
+        // Check the right neighbor connection, kill and restart if needed.
+        // Check if data needs to be shifted with the right node
+        // This mutex is locked for a long time  TODO break this up?
+        std::lock_guard<std::mutex> rMtxLG(_rightMtx);
+        LOGS(_log, LOG_LVL_INFO, "_monitor " << _ourId <<
+                                 " checking right neighbor " << _neighborRight.getId());
+        if (_neighborRight.getId() != 0) {
+            try {
+                if (not _neighborRight.getEstablished()) {
+                    LOGS(_log, LOG_LVL_INFO, "_monitor " << _ourId << " trying to connect");
+                    auto nAddr = _neighborRight.getAddressTcp();
+                    if (nAddr.ip == "") {
+                        // look up the network address for the rightNeighbor
+                        WWorkerListItem::Ptr nWorker =
+                            _wWorkerList->getWorkerNamed(_neighborRight.getId());
+                        if (nWorker != nullptr) {
+                            auto addr = nWorker->getAddressTcp();
+                            auto addrUdp = nWorker->getAddressUdp();
+                            LOGS(_log, LOG_LVL_INFO, "_monitor neighbor right " <<
+                                 _neighborRight.getId() << " T=" << addr << " U=" << addrUdp);
+                            _neighborRight.setAddressTcp(addr);
+                            _neighborRight.setAddressUdp(addrUdp);
+                        }
+                    }
+
+                    LOGS(_log, LOG_LVL_INFO, "_monitor trying to establish TCP connection with " <<
+                            _neighborRight.getId() << " " << _neighborRight.getAddressTcp());
+                    _rightConnect(rMtxLG); // calls _determineRange() while establishing connection
+                } else {
+                    LOGS(_log, LOG_LVL_INFO, "_monitor " << _ourId << " getting range info");
+                    if (_determineRange()) {
```

Comment:
This makes a lot of remote calls, while holding the `_rightMtx`. Seems likely to become problematic.

Reply:
I don't like it either, hence the TODO break this up? comment above. I haven't thought of a good improvement yet. This function and the _masterDisable() function are currently the only ones using the mutex, and that one isn't really used yet. So, not urgent, but icky.